### ISSUE TITLE
(WIP) Support regex in matchers

### ISSUE THREAD

**Issue by asafkorem** [wix/Detox#3366]
#### Support regular expression in Detox matchers
Specifically in `by.text()`, `by.label()`.
This can be useful since text and labels may include dynamic content (e.g. date).

Also, I suggest to change the current API, from:
https://github.com/wix/Detox/blob/246317d3ee5b7587bb966c78204b5def5b72dfaf/detox/index.d.ts#L804

To:
```ts
text(text: string | RegExp): NativeMatcher;
```

---

#### Usage example

This code will find an element with the text `"The year is 2023"`, if exists:

```js
const someElement = element(by.text(/The year is: \d{4}/));
```

While this code won't:

```js
const someElement = element(by.text('The year is: \d{4}'));
```

**Comment by noomorph** [comment]
Why not dynamically recognize `instanceof Regexp` versus string?

**Comment by asafkorem** [comment]
> Why not dynamically recognize `instanceof RegExp` versus string?

Sounds great, let's do it this way.

**Comment by noomorph** [comment]
@asafkorem the only pitfall I see is that we might need to standardize RegExp mechanisms between iOS/Android/JS...

<table>
<td>
<img width=300 src=https://user-images.githubusercontent.com/1962469/166209143-90048c25-cf1b-4c5e-b613-b55fc574c84d.png>
</td>
<td>
<img src=https://user-images.githubusercontent.com/1962469/166209174-597d49b0-9746-4d5a-8341-b0510405d892.png>
</td>
</table>

**Comment by d4vidi** [comment]
@noomorph that is spot on. We've somehow manged to overcome that in the context of the URL's blacklist - we could revisit and hopefully replicate.

**Comment by d4vidi** [comment]
Nevertheless I'd like this kind of any API to come with a warning of some sort, that would explains that ideally, whichever text we trying to match should be deterministic to the point it can be predetermined down to the character.
Of course, this is not always the case in unmocked tests, but generally this should be the preferred situation.

**Comment by boxabirds** [comment]
This is super useful when working with iOS elements because you can't seem to expose testIDs (or equivalent) so I'm using `accessibilityLabel` to perform the function of both an identifier and additional state.

**Comment by mariushorvat** [comment]
Is someone working on this issue?

**Comment by noomorph** [comment]
Not really, @asafkorem 's been busy last months trying to release Detox 21 with XCUITest.

@asafkorem would you mind to take this issue to the final mile? Looks like there is demand for it, 10*👍 already

**Comment by d4vidi** [comment]
Side note (because I couldn't find it easily beforehand): The work that's been put into see is available in this [old branch](https://github.com/wix/Detox/tree/optimize-app-fix-regression)

**Comment by asafkorem** [comment]
This feature was released in [version 20.11.0](https://github.com/wix/Detox/releases/tag/20.11.0), thanks to @andrius-girdzius!

**Comment by nriccar** [comment]
iOS Detox always anchors your pattern (^…$), so if you want to match extra text around your string you still have to wrap it in wildcards. 

Function that works:
`
/**
 * Returns a Detox element matcher that will look for your `text`
 * anywhere inside the node’s text (case-insensitive).
 *
 * @param  {string} needle       The dynamic string you want to match
 * @param  {object} [opts]
 * @param  {boolean} [opts.startsWith=false]  If true, only anchor at the start
 * @param  {boolean} [opts.endsWith=false]    If true, only anchor at the end
 * @return {Detox.IndexableNativeElement}             A fully-formed `element(by.text(…))`
 */

export const elementByRegex = (
  text: string,
  opts: {
    startsWith?: boolean
    endsWith?: boolean
  } = {
    startsWith: false,
    endsWith: false
  }
): Detox.IndexableNativeElement => {
  const { startsWith, endsWith } = opts

  // escape any regex-special characters in the dynamic string
  const safe = text.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')

  // build your pattern
  const prefix = startsWith ? '^' : '.*'
  const suffix = endsWith ? '$' : '.*'
  const pattern = `${prefix}${safe}${suffix}`

  const regex = new RegExp(pattern, 'i')
  return element(by.text(regex))
}
`

this will allow partial matches, like:
`
elementByRegex('test')
`

having: 
`
<Text>this is a test text</Text>
`
